### PR TITLE
PanelMenu: Fixes panel submenu not being accessible for panels close to the right edge of the screen

### LINKED ISSUE #15692: Panel Menu More... is shown off screen / window when panel is small and to the far right
Read before posting: 

- Questions should be posted to https://community.grafana.com. Please search there and here on GitHub for similar issues before creating a new issue. 
- Checkout FAQ: https://community.grafana.com/c/howto/faq
- Checkout How to troubleshoot metric query issues: https://community.grafana.com/t/how-to-troubleshoot-metric-query-issues/50

Please include this information:
### What Grafana version are you using?
v5.4.2
### What datasource are you using?
prometheus
### What OS are you running grafana on?
RedHat
### What did you do?
Devops
### What was the expected result?
expencting the full menu
### What happened instead?
Menu is not showing
### If related to metric query / data viz:
Its related to Grafana GUI
If I create  6 single stat per row the 6th one (that is last  in the right) showing only 2 character of the sub menu when More... is clicked from the menu
### Include raw network request & response: get by opening Chrome Dev Tools (F12, Ctrl+Shift+I on windows, Cmd+Opt+I on Mac), go the network tab.
 NA

### LINKED ISSUE #23095: Panel menu's submenu is not visible when panel is on the right side of the screen
**What happened**: If a Table's on the right side of the screen, the sub-contents of the "More..." menu cannot be accessed.

**How to reproduce it (as minimally and precisely as possible)**: Create a narrow Table to the right side of the screen and try to export to CSV. 

**Environment**:
- Grafana version: v6.7.1 (ca6d08d5cb)
- OS Grafana is installed on: CentOS
- User OS & Browser: Chrome 80.0.3987.100 (64 bit)

image

### LINKED ISSUE #25549: Panel: Sub Menu option not displayed when the panel has long title and sitting at the left hand side
**What happened**:
Observe the panel 'client side full page load' at the left bottom corner.  The panel menu is displayed okay. However, clicking on 'More' does not display the expected sub-menu.

image

**What you expected to happen**:
Sub-menu should be displayed on the right hand side, if there is no real estate available to display on the left hand side. 

**How to reproduce it (as minimally and precisely as possible)**:
- Add a panel with long title on the left hand side of the dashboard.
- Open Panel Menu, goto More'.  'Copy, Duplicate' options are not visible.

**Anything else we need to know?**:

**Environment**:
- Grafana version: 7.0.3
- Data source type & version: NA
- OS Grafana is installed on: NA
- User OS & Browser: Mac/Chrome
- Grafana plugins: NA
- Others: NA

### PR DESCRIPTION
**Pull request by torkelo**
This fixes a problem that has existed for an embarrassingly long time. Started exploring using popper for this, or creating our own dropdown component based on rc-dropdown but that would also mean creating components for Menu etc. and wanted something that we can include in a patch release so tried to adopt the fix we did for dashboard link dropdowns: https://github.com/grafana/grafana/pull/28330/  and discovered that that fix did not work, always made it pull left and did not check window boundary.  

Tried to share logic between the two but ended duplicating the function as they need to check different properties (left vs right) so ended duplicated in the end. Should be fine as this code will go away when we add dropdown & menu components. 

Fixes #25549
Fixes #15692
Fixes #23095

Propper fix fo #27900

### CODE REVIEW

File: public/app/features/dashboard/components/SubMenu/DashboardLinksDashboard.tsx [hunk 1]
```diff
@@ -62,7 +62,7 @@ export class DashboardLinksDashboard extends PureComponent<Props, State> {
           resolvedLinks.map((resolvedLink, index) => {
             const linkElement = (
               <a
-                className="gf-form-label"
+                className="gf-form-label && gf-form-label--dashlink"
```
Comment (dprokop) [inline public/app/features/dashboard/components/SubMenu/DashboardLinksDashboard.tsx:64]: What is this syntax?

File: public/sass/components/_dropdown.scss [hunk 2]
```diff
@@ -277,16 +278,21 @@
...
   // Undo the float
   // Yes, this is awkward since .pull-left adds a float, but it sticks to our conventions elsewhere.
-  float: none;
+  float: none !important;
 
   // Positioning the submenu
   > .dropdown-menu {
     left: -100%;
-    margin-left: 10px;
+    width: 100%;
```
Comment (torkelo) [inline public/sass/components/_dropdown.scss:286]: this was the one that was so hard to figure out, to make the submenu left: -100% positioning work it needs to be as wide as the parent

File: public/app/features/dashboard/dashgrid/PanelHeader/PanelHeaderMenuItem.tsx [hunk 3]
```diff
@@ -44,3 +46,22 @@ export const PanelHeaderMenuItem: FC<Props & PanelMenuItem> = props => {
     </li>
   );
 };
+
+export function getDropdownLocationCssClass(element: HTMLElement | null) {
```
Comment (dprokop) [inline public/app/features/dashboard/dashgrid/PanelHeader/PanelHeaderMenuItem.tsx:50]: Ah, we do similar things across multiple places. Think we finally need to create some solid abstraction around positioning elements.

**Review by dprokop — APPROVED** [review]
Surprisingly quite a lot of code! But it looks good, just one question about syntax that I see for the first time.

File: public/app/features/dashboard/components/SubMenu/DashboardLinksDashboard.tsx [hunk 4]
```diff
@@ -62,7 +61,7 @@ export class DashboardLinksDashboard extends PureComponent<Props, State> {
...
```
Comment (jackw) [inline public/app/features/dashboard/components/SubMenu/DashboardLinksDashboard.tsx:64]: Is this `&&` a typo?

Comment (dprokop) [reply public/app/features/dashboard/components/SubMenu/DashboardLinksDashboard.tsx:64]: guess it's "too much react" ;)

File: public/app/features/dashboard/components/SubMenu/DashboardLinksDashboard.tsx (hunk 1 above)
Comment (torkelo) [reply public/app/features/dashboard/components/SubMenu/DashboardLinksDashboard.tsx:64]: buggy brain syntax! well spotted

File: public/app/features/dashboard/dashgrid/PanelHeader/PanelHeaderMenuItem.tsx (hunk 3 above)
Comment (torkelo) [reply public/app/features/dashboard/dashgrid/PanelHeader/PanelHeaderMenuItem.tsx:50]: yea, I know I tried to but could not share it in the end

File: public/app/features/dashboard/components/SubMenu/DashboardLinksDashboard.tsx (hunk 4 above)
Comment (torkelo) [reply public/app/features/dashboard/components/SubMenu/DashboardLinksDashboard.tsx:64]: yes

File: public/app/features/dashboard/dashgrid/PanelHeader/PanelHeaderMenuItem.tsx (hunk 3 above)
Comment (dprokop) [reply public/app/features/dashboard/dashgrid/PanelHeader/PanelHeaderMenuItem.tsx:50]: No worries, I think we can have that as a separate task as we need to unify this across dropdowns, selects etc

File: public/sass/components/_dropdown.scss (hunk 2 above)
Comment (jackw) [reply public/sass/components/_dropdown.scss:286]: Could using something like `transform: translateX(-100%);` achieve the same result without locking down the width?

**Review by jackw — APPROVED** [review]
LGTM! 👏

File: public/sass/components/_dropdown.scss (hunk 2 above)
Comment (torkelo) [reply public/sass/components/_dropdown.scss:286]: Tested this now but did not work

**Pull request merged**